### PR TITLE
Feature/recursive videos to poses

### PR DESCRIPTION
**Pull request by cleong110**
Implemented a recursive option, solves #125 . Also added in an optional argument to specify the video extension to search for, though currently the only choice is .mp4. All behavior should be the same as before, as in if you don't use the -r argument it will just look in the current directory.

### CODE REVIEW

**Comment by cleong110** [comment]
Did a quick edit to the file to _SUPPORTED_VIDEO_FORMATS to see how easy it would be to switch extensions, and it successfully found and ran on .webm files as expected. But .webm support is a separate issue.  https://github.com/sign-language-processing/pose/issues/127

File: docs/specs/v0.1.md [hunk 1]
```diff
@@ -24,10 +24,16 @@
 \[`unsigned short` Green]
 \[`unsigned short` Blue]
 
+
+[Testing comment syntax from https://stackoverflow.com/a/32190021]: #
```
Comment (AmitMY) [inline docs/specs/v0.1.md:28]: why this test?

File: docs/specs/v0.1.md [hunk 2]
```diff
@@ -24,10 +24,16 @@
...
+[Testing comment syntax from https://stackoverflow.com/a/32190021]: # 
+
 # Body
-\[`unsined short` FPS]
-\[`unsined short` Number of frames] # THIS IS A PROBLEM
-\[`unsined short` Number of people]
+\[`unsigned short` FPS]
+
+
+[THIS IS A PROBLEM]: #
```
Comment (AmitMY) [inline docs/specs/v0.1.md:34]: yes. this is fixed in v0.2

File: src/python/pose_format/bin/directory.py [hunk 3]
```diff
@@ -1,45 +1,42 @@
...
-    pose_files = {removesuffix(f, ".pose") for f in all_files if f.endswith(".pose")}
     missing_pose_files = []
 
-    for mp4_file in mp4_files:
-        base_name = removesuffix(mp4_file, ".mp4")
-        if base_name not in pose_files:
-            missing_pose_files.append(os.path.join(directory, mp4_file))
-
-    return sorted(missing_pose_files)
+    for vid_file in vid_files:
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:17]: it is much faster to check all of the pose files
```py
glob_method = getattr (directory, 'rglob' if recursive else 'glob')
video_files = glob_method(f"*{video_suffix}")
pose_files = glob_method(f"*.pose")
```

File: src/python/pose_format/bin/directory.py [hunk 4]
```diff
@@ -1,45 +1,42 @@
...
-    parser.add_argument('--format',
+    parser.add_argument('-f','--format',
                         choices=['mediapipe'],
                         default='mediapipe',
                         type=str,
                         help='type of pose estimation to use')
-    parser.add_argument("--directory", type=str, required=True)
+    parser.add_argument("-d","--directory", type=Path, required=True, help="Directory to search for videos in")
+    parser.add_argument("-r", "--recursive", action="store_true", help="Whether to search for videos recursively")
+    parser.add_argument("--video_suffix", type=str,
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:32]: write `video-suffix` (dash, no underscore)

File: src/python/pose_format/bin/directory.py [hunk 5]
```diff
@@ -1,45 +1,42 @@
...
                         choices=['mediapipe'],
                         default='mediapipe',
                         type=str,
                         help='type of pose estimation to use')
-    parser.add_argument("--directory", type=str, required=True)
+    parser.add_argument("-d","--directory", type=Path, required=True, help="Directory to search for videos in")
+    parser.add_argument("-r", "--recursive", action="store_true", help="Whether to search for videos recursively")
+    parser.add_argument("--video_suffix", type=str, 
+                        choices=_SUPPORTED_VIDEO_FORMATS,
+                        default=".mp4", help="Video extension to search for")
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:34]: how about by default, we just go through all formats? something like `{'.mp4', '.mov', '.avi', '.mkv', '.flv', '.wmv', '.webm'}`

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: docs/specs/v0.1.md (hunk 1 above)
Comment (cleong110) [reply docs/specs/v0.1.md:28]: Ah, leftover bit.

File: src/python/pose_format/bin/directory.py (hunk 5 above)
Comment (cleong110) [reply src/python/pose_format/bin/directory.py:34]: Sure, that makes sense to me

Comment (cleong110) [reply src/python/pose_format/bin/directory.py:34]: I was going for minimal behavior changes in this particular PR, but I do think just checking them all seems best

**Comment by cleong110** [comment]
Updated code to check against list of pose files as requested, and to iterate over all video suffixes. Realized this could result in foo.mp4, foo.webm both being posed to foo.pose, so added an option to keep original video suffix, and added a "this will cause name collisions do you want to keep going" check

**Comment by cleong110** [comment]
Also added a try/except checking for ValueError, so that if one of the video files is corrupted, the utility will keep going.

Also added a check for existing .pose files, in which case it will not re-run/overwrite.

File: docs/specs/v0.1.md [hunk 6]
```diff
@@ -24,10 +24,13 @@
...
 
+
 # Body
-\[`unsined short` FPS]
-\[`unsined short` Number of frames] # THIS IS A PROBLEM
-\[`unsined short` Number of people]
+\[`unsigned short` FPS]
+
+
+\[`unsigned short` Number of frames]
```
Comment (AmitMY) [inline docs/specs/v0.1.md:32]: no need to modify the spec

File: src/python/pose_format/bin/directory.py [hunk 7]
```diff
@@ -1,45 +1,171 @@
...
-import os
-
+from pathlib import Path
 from pose_format.bin.pose_estimation import pose_video, parse_additional_config
+from typing import List
+import logging
 from tqdm import tqdm
 
+# Note: untested other than .mp4. Support for .webm may have issues: https://github.com/sign-language-processing/pose/pull/126
+_SUPPORTED_VIDEO_FORMATS = [".mp4", ".mov", ".avi", ".mkv", ".flv", ".wmv", ".webm"]
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:9]: not a private variable, can be just uppercase, no initial underscore

File: src/python/pose_format/bin/directory.py [hunk 8]
```diff
@@ -1,45 +1,171 @@
...
+
+    # Prevents the common gotcha with mutable default arg lists: 
+    # https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments
+    if video_suffixes is None:
+        video_suffixes = _SUPPORTED_VIDEO_FORMATS
+
+    glob_method = getattr(directory, "rglob" if recursive else "glob")
+    all_files = list(glob_method(f"*"))
+    video_files = [path for path in all_files if path.suffix in video_suffixes]
+    pose_files = [path for path in all_files if path.suffix == ".pose"]
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:48]: pose files should be a `set` to make the check fast

File: src/python/pose_format/bin/directory.py [hunk 9]
```diff
@@ -1,45 +1,171 @@
...
-def removesuffix(text: str, suffix: str):
-	if text.endswith(suffix):
-		return text[:-len(suffix)]
-	else:
-		return text
+    videos_with_missing_pose_files = []
 
+    for vid_path in video_files:
+        if (
+            not get_corresponding_pose_path(
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:54]: the if statement is ugly. use a variable `corresponding_pose_path = get....` and then check

File: src/python/pose_format/bin/directory.py [hunk 10]
```diff
@@ -1,45 +1,171 @@
...
+        or replace (foo.mp4->foo.pose). Defaults to replace.
+
+    Returns
+    -------
+    Path
+        pathlib Path
+    """
+    if keep_video_suffixes:
+        return video_path.with_name(f"{video_path.name}.pose")
+    else:
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:85]: no need for `else`

File: src/python/pose_format/bin/directory.py [hunk 11]
```diff
@@ -1,45 +1,171 @@
...
+    videos_with_missing_pose_files = find_videos_with_missing_pose_files(
+        args.directory,
+        video_suffixes=args.video_suffixes,
+        recursive=args.recursive,
+        keep_video_suffixes=args.keep_video_suffixes,
+    )
+
+    print(f"Found {len(videos_with_missing_pose_files)} videos missing pose files.")
+
+    pose_files_that_will_be_created = set(
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:140]: if you do `{... for ... in ...}` that is a set, no need to create a list and cast a set

File: src/python/pose_format/bin/directory.py [hunk 12]
```diff
@@ -1,45 +1,171 @@
...
-    for mp4_path in tqdm(missing_pose_files):
-        pose_file_name = removesuffix(mp4_path, ".mp4") + ".pose"
-        pose_video(mp4_path, pose_file_name, args.format, additional_config)
+    pose_with_no_errors = []
+
+    for vid_path in tqdm(videos_with_missing_pose_files):
+        try:
+            pose_path = get_corresponding_pose_path(
+                video_path=vid_path, keep_video_suffixes=args.keep_video_suffixes
+            )
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:162]: what's with all these line breaks? i feel like most of the lines would fit under 140 characters, but they are broken in most of the file

File: src/python/pose_format/bin/directory.py [hunk 13]
```diff
@@ -1,45 +1,171 @@
...
+            if pose_path.is_file():
+                print(f"Skipping {vid_path}, corresponding .pose file already created.")
+                continue
+            pose_video(vid_path, pose_path, args.format, additional_config)
+            pose_with_no_errors.append(vid_path)
```
Comment (AmitMY) [inline src/python/pose_format/bin/directory.py:167]: if this is basically a counter, just count - `+=1` no need to create another array

**Review by AmitMY — CHANGES_REQUESTED** [review]
much better. still minor changes

File: src/python/pose_format/bin/directory.py (hunk 11 above)
Comment (cleong110) [reply src/python/pose_format/bin/directory.py:140]: Huh! TIL

File: src/python/pose_format/bin/directory.py (hunk 12 above)
Comment (cleong110) [reply src/python/pose_format/bin/directory.py:162]: I ran `black` on it, it must have been just over the threshold.

File: src/python/pose_format/bin/directory.py (hunk 13 above)
Comment (cleong110) [reply src/python/pose_format/bin/directory.py:167]: I had thought to use it at a later date for more detailed logging/debugging, e.g. write out the specific ones that didn't work. But I can just add the list then when/if it's needed.

File: docs/specs/v0.1.md (hunk 6 above)
Comment (cleong110) [reply docs/specs/v0.1.md:32]: Should I put it back in?

File: src/python/pose_format/bin/directory.py (hunk 12 above)
Comment (cleong110) [reply src/python/pose_format/bin/directory.py]: Readjusted manually, then re-ran black with --line-length 140

File: docs/specs/v0.1.md (hunk 6 above)
Comment (AmitMY) [reply docs/specs/v0.1.md:32]: yeah i would revert the change to this file, no need to change it for this PR and there is just a comment removed

**Review by AmitMY — APPROVED** [review]
LGTM - just revert one change and i'll merge.
Thanks!

File: docs/specs/v0.1.md (hunk 6 above)
Comment (cleong110) [reply docs/specs/v0.1.md:33]: Done!

**Review by AmitMY — APPROVED** [review]

**Comment by J22Melody** [comment]
It is great! 

I will add a multiprocessing option to the script to use multiple CPUs. Currently, I need to run pose estimation on 1000+ hours of BSL data, and do not want to wait for too long ;)

**Comment by cleong110** [comment]
https://github.com/cleong110/ud-vlab/blob/main/data_munging/split_folder.py
https://github.com/cleong110/ud-vlab/blob/main/data_munging/run_videos_to_poses_on_multiple_dirs.sh

**Comment by J22Melody** [comment]
@cleong110 please share the repo with me!

**Comment by cleong110** [comment]
Oh silly me! The two files in question:
[run_videos_to_poses_on_multiple_dirs.zip](https://github.com/user-attachments/files/18027016/run_videos_to_poses_on_multiple_dirs.zip)

**Pull request merged**